### DISCUSSION TITLE
Household create 5.1 set

### DISCUSSION THREAD

**Discussion by Ptrvh** [hklages/node-red-contrib-sonos-plus#238] [Ideas]
I love the Sonos Plus functionality and recently I used it to do the following:

Pair a Symfonisk Lamp and a Sonos One as a stereo pair. By default this cannot be done in the Sonos app but with the Contrib-Sonos-Plus function you can. Great!

But. It would be really nice if I could add surround speakers to my Arc and Sub from two different speakers (the Symfonisk Lamp and a Sonos One in this case). Or, also possible maybe: 2 Sonos Roam's (not possible by default).

Is it possible to generate a function to do this?

**Comment by Ptrvh** [comment]
Found this in the Sonos api documentation: 

<h3 id="addhtsatellite" style="font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-tap-highlight-color: rgba(26, 26, 26, 0.3); -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; margin-top: 1.5em; margin-bottom: 0.25em; font-weight: 500; line-height: 1.25; color: rgb(39, 38, 43); position: relative; font-family: system-ui, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, sans-serif; -webkit-text-size-adjust: 100%; font-size: 16px !important;">AddHTSatellite<span class="Apple-converted-space"> </span></h3><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-tap-highlight-color: rgba(26, 26, 26, 0.3); -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; margin-top: 0px; margin-bottom: 1em; caret-color: rgb(92, 89, 98); color: rgb(92, 89, 98); font-family: system-ui, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 14px; -webkit-text-size-adjust: 100%;">Action body:</p><div class="language-xml highlighter-rouge" style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-tap-highlight-color: rgba(26, 26, 26, 0.3); -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; margin-top: 0px; padding: 0.75rem; margin-bottom: 0.75rem; overflow-x: auto; background-color: rgb(245, 246, 250); border-radius: 4px; box-shadow: none; -webkit-overflow-scrolling: touch; caret-color: rgb(92, 89, 98); color: rgb(92, 89, 98); font-family: system-ui, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 14px; -webkit-text-size-adjust: 100%;"><div class="highlight" style="box-sizing: border-box; margin: 0px; padding: 0px; border: 0px;"><pre class="highlight" style="box-sizing: border-box; font-family: monospace, monospace; font-size: 1em; margin: 0px; padding: 0px; border: 0px;"><code style="box-sizing: border-box; font-family: SFMono-Regular, Menlo, Consolas, monospace; font-size: 0.75em; line-height: 1.4; padding: 0px; font-weight: 400; background-color: rgb(245, 246, 250); border: 0px; border-radius: 4px; margin: 0px;"><span class="nt" style="box-sizing: border-box; color: navy;">&lt;u:AddHTSatellite</span> <span class="na" style="box-sizing: border-box; color: teal;">xmlns:u=</span><span class="s" style="box-sizing: border-box; color: rgb(221, 17, 68);">"urn:schemas-upnp-org:service:DeviceProperties:1"</span><span class="nt" style="box-sizing: border-box; color: navy;">&gt;</span>
  <span class="nt" style="box-sizing: border-box; color: navy;">&lt;HTSatChanMapSet&gt;</span>string<span class="nt" style="box-sizing: border-box; color: navy;">&lt;/HTSatChanMapSet&gt;</span>
<span class="nt" style="box-sizing: border-box; color: navy;">&lt;/u:AddHTSatellite&gt;</span>
</code></pre></div></div><p style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-tap-highlight-color: rgba(26, 26, 26, 0.3); -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; margin-top: 1em; margin-bottom: 1em; caret-color: rgb(92, 89, 98); color: rgb(92, 89, 98); font-family: system-ui, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 14px; -webkit-text-size-adjust: 100%;">Inputs:</p><div class="table-wrapper" style="font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-tap-highlight-color: rgba(26, 26, 26, 0.3); -webkit-text-stroke-width: 0px; text-decoration: none; box-sizing: border-box; margin-top: 0.5em; display: block; width: 347px; max-width: 100%; margin-bottom: 1.5rem; overflow-x: auto; border-radius: 4px; box-shadow: rgba(0, 0, 0, 0.12) 0px 1px 2px, rgba(0, 0, 0, 0.08) 0px 3px 10px; caret-color: rgb(92, 89, 98); color: rgb(92, 89, 98); font-family: system-ui, -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 14px; -webkit-text-size-adjust: 100%;">

parameter | type | description
-- | -- | --
HTSatChanMapSet | string

</div>

**Comment by hklages** [comment]
> Pair a Symfonisk Lamp and a Sonos One as a stereo pair. By default this cannot be done in the Sonos app but with the Contrib-Sonos-Plus function you can. Great!

Yes. That was also good for me because I could test it with Play3 and play1 - I don't have 2 of same type. 

> But. It would be really nice if I could add surround speakers to my Arc and Sub from two different speakers (the Symfonisk Lamp and a Sonos One in this case). Or, also possible maybe: 2 Sonos Roam's (not possible by default).

How do yo do that in the app? Are surround speakers the same as "satellites"? How many can you add? 2, one for left, one for right channel? 

Concrete: We should try to add the Symfonisk Lamp and a Sonos One to your Sonos Arc as satellites? And I guess we also need a command to "remove" them. 

I found a string, what might show the syntax: coordinator, right speaker, left speaker and subwofer? 
Could that be? 

`HTSatChanMapSet="RINCON_000PPP1400:LF,RF;RINCON_000RRR1400:RR;RINCON_000SSS1400:LR;RINCON_000QQQ1400:SW"`

**Reply by Ptrvh** [reply]
In the app you go to System and select the Home Theater product, in my case the Arc+Sub. There you can add Surround speakers (this is the same as Satellites I think indeed). You can add in the App 2 the same speakers but it is limited to certain speakers. You cannot add the Roam for example. 
![add satellite](https://user-images.githubusercontent.com/22859779/211095340-071a00b3-2df4-4480-8921-db2121085308.jpg)

Since you can stereo pair with the command different speakers I thought it should be possible to add two different speakers as a sattelite as well. Or maybe 2 Roam's.

Concrete indeed: We should try to add the Symfonisk Lamp and a Sonos One to your Sonos Arc as satellites? And we also need a command to "remove" them.

The String you found is related indeed, In my second post I also shared something which I found in the API documentation.


I don't have the knowledge to program code (I can only copy paste and adjust :) ). I have plenty of Sonos products to test with.

**Reply by Ptrvh** [reply]
> `HTSatChanMapSet="RINCON_000PPP1400:LF,RF;RINCON_000RRR1400:RR;RINCON_000SSS1400:LR;RINCON_000QQQ1400:SW"`

I Think the first one is the Arc (LF and RF) suppose that are Left front and Right Front and this is one product. Then we have the Right Rear and then the Left Rear. Last one is the Sub Woofer.

**Reply by hklages** [reply]
Are you familiar with NodeRED and editing JSON structures?  
For a first test I am going to prepare a flow sending the raw data to the player.  You need to add the IP address and the "RINCON". Maybe Saturday or Sunday.

**Reply by Ptrvh** [reply]
I think I will manage that. IP's and RinconID's are available. Looking forward to it.

**Comment by hklages** [comment]
Hi. Here is the flow for testing. I was able to add the Play:1 to a SonosBeam (with an error message) and to remove it. 
It was not possible to add a Play:1 and a Roam. 

The first path is only to test the hidden function - it should debug your household id. That should in any case work. 

`[{"id":"6767f287056b4fd5","type":"inject","z":"e31b5e93c94525f4","name":"household id","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"player.execute.action.v8","payload":"{\"endpoint\":\"/DeviceProperties/Control\",\"action\":\"GetHouseholdID\",\"inArgs\":{}}","payloadType":"json","x":1229,"y":66,"wires":[["ef8171cb1c0fb0b0"]]},{"id":"ef8171cb1c0fb0b0","type":"sonos-universal","z":"e31b5e93c94525f4","confignode":"8d44022c.adb65","command":"message","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1454,"y":138,"wires":[["cfd20d74b39f0b6d"]]},{"id":"cfd20d74b39f0b6d","type":"debug","z":"e31b5e93c94525f4","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":1611,"y":138,"wires":[]},{"id":"7593b5fe1d9fd449","type":"inject","z":"e31b5e93c94525f4","name":"Remove Satellite","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"player.execute.action.v8","payload":"{\"endpoint\":\"/DeviceProperties/Control\",\"action\":\"RemoveHTSatellite\",\"inArgs\":{\"SatRoomUUID\":\"RINCON_B8E9375831C001400\"}}","payloadType":"json","x":1240,"y":220,"wires":[["ef8171cb1c0fb0b0"]]},{"id":"a74be5a8d95fd1b9","type":"inject","z":"e31b5e93c94525f4","name":"Create 5.1","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"player.execute.action.v8","payload":"{\"endpoint\":\"/DeviceProperties/Control\",\"action\":\"AddHTSatellite\",\"inArgs\":{\"HTSatChanMapSet\":\"RINCON_949F3EC13B9901400:LF,RF;RINCON_B8E9375831C001400:LR\"}}","payloadType":"json","x":1220,"y":140,"wires":[["ef8171cb1c0fb0b0"]]},{"id":"77e1066ff2761b29","type":"comment","z":"e31b5e93c94525f4","name":"modify RINCONN","info":"RINCON... should be the \nname of the player to be removed!","x":1268,"y":190,"wires":[],"icon":"font-awesome/fa-long-arrow-down"},{"id":"02dc18f838ddf0ac","type":"comment","z":"e31b5e93c94525f4","name":"modify IP in config","info":"IP should be the address of\nthe Sonos BEAM or Playbar","x":1483,"y":108,"wires":[],"icon":"font-awesome/fa-long-arrow-down"},{"id":"4dbf5e53c8e033e8","type":"comment","z":"e31b5e93c94525f4","name":"modify RINCONNS","info":"Add your RINCONS, first the BEAM,  \nthen the other player\n\nRINCON_<beam>:LF,RF;RINCON_<right rear>:RR;RINCON_<left rear>:LR; RINCON_<sub>:SW","x":1276,"y":110,"wires":[],"icon":"font-awesome/fa-long-arrow-down"},{"id":"8d44022c.adb65","type":"sonos-config","name":"your_tvplayer_dns","serialnum":"","ipaddress":"192.168.178.54"}]`

Please let me know your results. If it works I will create the commands in universal node.

**Comment by Ptrvh** [comment]
![sonos satellite](https://user-images.githubusercontent.com/22859779/211173049-46b6e614-7cd9-411c-aeab-ba5d06c1b36e.png)

:) Works perfect! Thanks

I tested it with:
- 2x One and Sub: Works perfect
- 1x One and 1x Symphonisk Lamp and Sub: Works perfect
- 1x Roam and 1x Roam SL and Sub: Error

Seems Roam really does not work.

Deleting sattelites does work one by one. I tried to reformat the JSON to delete them both in one command but did not succeed. Maybe you have an idea how to try?

JSON for adding Sattelite and SUB:
```
{
    "endpoint": "/DeviceProperties/Control",
    "action": "AddHTSatellite",
    "inArgs": {
        "HTSatChanMapSet": "RINCON_48A6B8B5614E01400:LF,RF;RINCON_38420B92ABE601400:RR;RINCON_7828CA042C8401400:LR;RINCON_542A1B108A6201400:SW"
    }
}
```

**Comment by hklages** [comment]
Great - and thanks for your feedback. The code is almost complete. I need to do some testing and will let you know, when the new release is available. 

The new command uses SONOS-Playernames instead of RINCONN_xxx and derive the RINCONN from the system. Hope thats easier for everyone. 

> Deleting sattelites does work one by one.

Right. That's why I added a new property "HTSatChanMapSet" to the group data. I can then use that, extract the uuids and do iterate inside the program. So you will only need one command "household.seperate.surroundsystem" to remove all satellites from the main player. 

2023-01-10 16:16: Picture delete as this does not work anymore.

**Comment by hklages** [comment]
Hi again. So 6.5.0 is on the server. The Wiki/documentation will follow soon. Here is an example flow, showing the new properties -all are mandatory. All are set in the inject node. 

Does it work in your environment? 

2023-01-10 16:16: Picture and flow delete, as it does not work anymore

**Reply by Ptrvh** [reply]
Updated to 6.5.0 and tested the new possibilities:

- Separate: Works very fast. But, with the current setup you separate everything, also the sub. With the test flow you could delete only the satellites. In my situation I only want to remove the satellites and leave the sub connected to the ARC. Maybe it is possible to set next to the player which you want to separate from also what you want to separate?

create: works perfect. Some remarkts:
- I doubted about the name of the Sub. Normally it does not have a name because it cannot be accessed as a speaker in the Sonos app. But if you joined if before with a speaker it has the name of the speaker it seams *after you separate it via the flow.. So it works but it's a bit strange.
- If you already have a 5.0 set and you only want to join the sub it does not work if you leave the LR and RR empty. If you fill them it gives an error because they are not found (since they are already linked).
- The same for a 3.1 set. If you want to add the satellites you have the same behavior as with adding the Sub.
I did not test yet if this was also with the JSON injection which you created before. Probably it does.
edit: Since it does not search the player but it uses the Rincon it is probably overwritten which makes it work.

But if you know these are the limitations it is a very nice solution which does what is needs to do!

Thank you very much, nice Job!

**Comment by hklages** [comment]
Excellent feedback - thanks. 

So I am going to rename the two commands
 - create.surroundsystem to add.satellites and it only adds the 2 satellites (not adding a sub)
 - separate.surroundsystem to remove.satellites and it only removes one or both satellites (but not the sub)
and in addition there is a new command
- add.subwoofer 

Removing the subwoofer can be done in the app - that will handle the naming correctly - I hope.

**Comment by hklages** [comment]
6.6.0 with the following commands: 

![sub satellites](https://user-images.githubusercontent.com/17273119/211589153-a055728b-2164-4a66-aa20-2c84f030419d.jpg)

and example flow: 

`
[{"id":"7d647adc160207f3","type":"inject","z":"fcb24b59.997048","name":"set sub","props":[{"p":"payload"},{"p":"playerNameSubwoofer","v":"SonosBad","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"SonosWohnzimmer","payloadType":"str","x":1370,"y":198,"wires":[["76ec73da146827ef"]]},{"id":"76ec73da146827ef","type":"sonos-universal","z":"fcb24b59.997048","confignode":"88e9fa70.4a53e8","command":"household.add.subwoofer","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1590,"y":198,"wires":[[]]},{"id":"ed4e36d19f0f2ce5","type":"inject","z":"fcb24b59.997048","name":"set sat","props":[{"p":"payload"},{"p":"playerNameLeftRear","v":"SonosRoam","vt":"str"},{"p":"playerNameRightRear","v":"SonosWintergarten","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"SonosWohnzimmer","payloadType":"str","x":1370,"y":244,"wires":[["d3caabc13cd3f925"]]},{"id":"d3caabc13cd3f925","type":"sonos-universal","z":"fcb24b59.997048","confignode":"88e9fa70.4a53e8","command":"household.add.satellites","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1590,"y":244,"wires":[[]]},{"id":"0e74abc68cb15af6","type":"sonos-universal","z":"fcb24b59.997048","confignode":"88e9fa70.4a53e8","command":"household.remove.satellites","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1600,"y":290,"wires":[[]]},{"id":"8d7df2cf4997ab3e","type":"inject","z":"fcb24b59.997048","name":"remove sat","props":[{"p":"payload"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"SonosWohnzimmer","payloadType":"str","x":1380,"y":290,"wires":[["0e74abc68cb15af6"]]},{"id":"88e9fa70.4a53e8","type":"sonos-config","name":"your_player_dns","serialnum":"","ipaddress":"sonosplay5.fritz.box"}]
`

**Comment by hklages** [comment]
> I doubted about the name of the Sub. Normally it does not have a name because it cannot be accessed as a speaker in the Sonos app. But if you joined if before with a speaker it has the name of the speaker it seams *after you separate it via the flow.. So it works but it's a bit strange.

I am not familiar with Subwoofer. They dont have a playername - right? 
So the add.subwoofer needs another identifier. What would be the best: the IP-address or the RINCONN_xxxx. Which one can you see in the SONOS app.

**Reply by Ptrvh** [reply]
> I am not familiar with Subwoofer. They dont have a playername - right?

Correct.

In the App you can find information about the system and there the IP is shown. The RINCON is only shown if you go to one of the Players via an URL like: http://ipaddress:1400/support/review

So I would suggest to use IP instead.

I will install 6.6.0 tonight and test it.

**Comment by hklages** [comment]
6.6.1 uses for subwoofer now "ipSubwoofer". 

`[{"id":"7d647adc160207f3","type":"inject","z":"fcb24b59.997048","name":"set sub","props":[{"p":"payload"},{"p":"ipSubwoofer","v":"192.168.178.52","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"SonosWohnzimmer","payloadType":"str","x":1370,"y":198,"wires":[["76ec73da146827ef"]]},{"id":"76ec73da146827ef","type":"sonos-universal","z":"fcb24b59.997048","confignode":"88e9fa70.4a53e8","command":"household.add.subwoofer","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1590,"y":198,"wires":[[]]},{"id":"ed4e36d19f0f2ce5","type":"inject","z":"fcb24b59.997048","name":"set sat","props":[{"p":"payload"},{"p":"playerNameLeftRear","v":"SonosRoam","vt":"str"},{"p":"playerNameRightRear","v":"SonosWintergarten","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"SonosWohnzimmer","payloadType":"str","x":1370,"y":244,"wires":[["d3caabc13cd3f925"]]},{"id":"d3caabc13cd3f925","type":"sonos-universal","z":"fcb24b59.997048","confignode":"88e9fa70.4a53e8","command":"household.add.satellites","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1590,"y":244,"wires":[[]]},{"id":"0e74abc68cb15af6","type":"sonos-universal","z":"fcb24b59.997048","confignode":"88e9fa70.4a53e8","command":"household.remove.satellites","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1600,"y":290,"wires":[[]]},{"id":"8d7df2cf4997ab3e","type":"inject","z":"fcb24b59.997048","name":"remove sat","props":[{"p":"payload"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"SonosWohnzimmer","payloadType":"str","x":1380,"y":290,"wires":[["0e74abc68cb15af6"]]},{"id":"88e9fa70.4a53e8","type":"sonos-config","name":"your_player_dns","serialnum":"","ipaddress":"sonosplay5.fritz.box"}]`

**Reply by Ptrvh** [reply]
Tested it.

1. Deleted Sub from system with hard reset. After that added the Sub to see the name. It is named: Sub...
2. Linked the Sub to the Arc. Removed it from Arc (all via Sonos App). Now it is available as Sub again.. so it means that the name it had previous was strange :).

Picture after I removed it:
<img width="80" alt="image" src="https://user-images.githubusercontent.com/22859779/212192502-5462c30b-0396-4507-b46a-3a48eedab080.png">

Next: added the Sub via IP in node red with the new function. Takes some time but works.
Removed Satellites via Node Red: works
Removed Sub via Sonos App: works and name stays correct
Added Sub to one of the removed satellites (this means I did not use the function to add the sub to a 5.1 set but to a single speaker) via node red: Error. The function only works with the 5.0 setup (a set where Satellites are already active).

error:
`"household.add.subwoofer:Sonos error on AddHTSatellite UPnPError 401 (Invalid action) :: Details: {"stack":"SonosError: Sonos error on AddHTSatellite UPnPError 401 (Invalid action)\n    at DevicePropertiesService.handleErrorResponse (/config/node-red/node_modules/@svrooij/sonos/lib/services/base-service.js:242:23)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at DevicePropertiesService.handleRequest (/config/node-red/node_modules/@svrooij/sonos/lib/services/base-service.js:199:16)\n    at DevicePropertiesService.SoapRequestWithBodyNoResponse (/config/node-red/node_modules/@svrooij/sonos/lib/services/base-service.js:133:16)\n    at DevicePropertiesService.AddHTSatellite (/config/node-red/node_modules/@svrooij/sonos/lib/services/device-properties.service.js:35:42)\n    at Object.householdAddSubwoofer [as household.add.subwoofer] (/config/node-red/node_modules/node-red-contrib-sonos-plus/src/sonos-universal.js:2447:5)","message":"Sonos error on AddHTSatell..."`

Added satellites via node red to ARC: works.
Added Sub via node red to 5.0 set (Arc + Satellites): works

Conclusion: The function works perfect to add a sub to a set with satellites. Function is not usable to just add a sub to a speakers.

I know, this was not the scope, but since now it implies you can add a sub to a Speaker I thought lets test this as well.

IP usage instead of Name works fine.

**Comment by Ptrvh** [comment]
Just did the following test since I noticed I did not.

start with a TV speaker which has nothing linked to it. In my case the Arc.

Add a Sub via node red: works
Add satellites to Arc + Sub: works.

conclusion: it works perfect as designed. 

The bonus feature to add the sub to a different speaker than a TV speaker is not possible with this function but actually this is not strange since it is a function to create a 5.1 set.

For me the function is well implemented and works perfect!

a future improvement could be to make a function to add a sub to any speaker but I don’t know if there is a function in the API for this.

**Comment by hklages** [comment]
... thanks for testing and the feedback. 

I will put a comment in the documentation (not possible to add sub)

Are you able to add a sub to that satellite with the the SONOS app? 
Maybe it is a bondedZone. When you use my household.get.group, what does it shows (chanel map, ht channel map) for that satellite?

If you like to further explore you could install a network sniffer and check what command SONOS is using - but that is not easy stuff. 
[Link to wireshark](https://github.com/SoCo/SoCo/wiki/Use-Wireshark-for-protocol-research)

**Reply by Ptrvh** [reply]
Maybe I'm not clear or I do not understand you.
- The current functionality works perfect. The function is to setup a Home Theater set, and that is what it does!

- a nice new feature would be: add a sub to a player via a function.

If you would like to add this and need help, no problem! If you want to leave it with the current functionality that's fine as well.
Let me know if you want to add additional functionality, if so I can help. But you need to explain a bit more in detail how because this is new for me.

**Comment by hklages** [comment]
I did understand you and I would like to investigate to add sub to player. 

Are you able to add a sub to one of your players (not to the 5.1 players) with the app? 
If yes please do it an send me the outcome of household.get.groups for that player - in particular the channel..., htchannel.... properties.

I hope that the pairing information is stored in on of them.

**Comment by Ptrvh** [comment]
OK,

Play:5 before adding Sub:
`[{"groupId":"RINCON_949F3E5092A801400:3060750834","uuid":"RINCON_949F3E5092A801400","urlObject":"http://192.168.2.31:1400/","playerName":"Keuken5","invisible":false,"channelMapSet":"","htSatChanMapSet":""}]`

Play:5 after adding Sub via Sonos App:
`[{"groupId":"RINCON_949F3E5092A801400:3060750834","uuid":"RINCON_949F3E5092A801400","urlObject":"http://192.168.2.31:1400/","playerName":"Keuken5","invisible":false,"channelMapSet":"RINCON_949F3E5092A801400:LF,RF;RINCON_542A1B108A6201400:SW,SW","htSatChanMapSet":""}]`

**Reply by Ptrvh** [reply]
Saw a pattern and tested it.

With the Create stereopair function we are able to do this. Used the first flow you created for injection and changed it to CreateStereoPair and set the ChannelMapSet as in the result after adding the Sub in the Sonos App, et Voila: it works.

```
{
    "endpoint": "/DeviceProperties/Control",
    "action": "CreateStereoPair",
    "inArgs": {
        "ChannelMapSet": "RINCON_949F3E5092A801400:LF,RF;RINCON_542A1B108A6201400:SW,SW"
    }
}
```

**Comment by Ptrvh** [comment]
Add PlayerIPSub to [this ](https://github.com/hklages/node-red-contrib-sonos-plus/wiki/A.1-Universal-Node#householdcreatestereopair)should do the trick with the ChannelMapSet above

And maybe remove could be done then as well.. should test this.

**Comment by hklages** [comment]
Great - you don't need me anymore :-)

Currently I am in Asia on vacation. I will create the new command after having returned: that means mid of February.

**Comment by hklages** [comment]
Hi - am back and working on the Sub for not HT devices. 

Could you be so kind and test the following flow. You have to enter your RINCON. Does it work? 

`[{"id":"ab6f181f2b673eaa","type":"debug","z":"e31b5e93c94525f4","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":1696,"y":560,"wires":[]},{"id":"5d17f168c660250d","type":"sonos-universal","z":"e31b5e93c94525f4","confignode":"88e9fa70.4a53e8","command":"message","state":"","stateType":"str","avoidCheckPlayerAvailability":false,"name":"","x":1540,"y":560,"wires":[["ab6f181f2b673eaa"]]},{"id":"2bab28fcde27ae0d","type":"inject","z":"e31b5e93c94525f4","name":"Add Sub (No HT 5.1)","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"player.execute.action.v8","payload":"{\"endpoint\":\"/DeviceProperties/Control\",\"action\":\"AddBondedZones\",\"inArgs\":{\"ChannelMapSet\":\"RINCON_5CAAFD00223601400:LF,RF;RINCON_000E58FE3AEA01400:SW\"}}","payloadType":"json","x":1327,"y":560,"wires":[["5d17f168c660250d"]]},{"id":"e521fe864682674f","type":"comment","z":"e31b5e93c94525f4","name":"modify RINCONN","info":"RINCON... should be the \nname of the player to be removed!","x":1340,"y":529,"wires":[],"icon":"font-awesome/fa-long-arrow-down"},{"id":"88e9fa70.4a53e8","type":"sonos-config","name":"your_player_dns","serialnum":"","ipaddress":"sonosplay5.fritz.box"}]`

**Comment by hklages** [comment]
Hi. 

I just published 6.7.0. In that release the add.subwoofer has an additional boolean property `isHTsystem`. This property is set to true as default but you can set it and then you should be able to add a sub to non Home Theater system.

Can you please verify. @Ptrvh

**Reply by Ptrvh** [reply]
Sorry, did not have time yet to test.. Screwed up my installation and had no time yet to replace with backup. Will check and let you know.

**Reply by Ptrvh** [reply]
A few months of delay.. sorry, summer is garden time. But I'm up and running again and gave it a try. Works perfect with the set of the HT boolean indeed! Great work.

**Reply by Ptrvh** [reply]
The funny thing is that if you create a flow to add it to a stereopair and a flow to add it to your HT you do not have to remove it first. I was searching for a funtion to remove the sub and we do not have one but you do not need it. If you add it to your HT when it is linked to a stereopair it just removes it and add it to the HT and visa versa. Perfect.

**Reply by Ptrvh** [reply]
hmm facing the problem that NodeRed says that the Subwoofer IP was not found.. Sonos App shows the IP and a Ping to the IP gives a direct response. Will test a bit more to find out what this can be.

EDIT: This might be caused because I run NodeRed in Docker.. had some problems before with accessing on IP level. But it is strange that it worked en then I just got an error.

**Comment by hklages** [comment]
In a Docker envirionment some things dont work 

Thanks for the feedback - I close this isssue now. Feel free to open a new one if you have other issues.
